### PR TITLE
mm: add invalid pid dump when malloc failed

### PR DESCRIPTION
**Pull request by anjiahao1**
## Summary
We can dump the memory that has exited but
has not been released.
## Impact
when malloc failed
## Testing
sim:nsh with testing_mm

### CODE REVIEW

**Review by xiaoxiang781216 — APPROVED** [review]

File: mm/mempool/mempool.c [hunk 1]
```diff
@@ -436,12 +437,15 @@ mempool_info_task(FAR struct mempool_s *pool,
       list_for_every_entry(&pool->alist, buf, struct mempool_backtrace_s,
                            node)
         {
-          if (buf->pid == dump->pid &&
-              buf->seqno >= dump->seqmin &&
-              buf->seqno <= dump->seqmax)
+          if (buf->pid == dump->pid ||
+              (dump->pid == MM_BACKTRACE_INVALID_PID &&
+                !nxsched_get_tcb(buf->pid)))
```
Comment (pkarashchenko) [inline mm/mempool/mempool.c:442]:
```suggestion
               nxsched_get_tcb(buf->pid) == NULL))
```

File: mm/mm_heap/mm_mallinfo.c [hunk 2]
```diff
@@ -98,16 +98,24 @@ static void mallinfo_task_handler(FAR struct mm_allocnode_s *node,
...
-#endif
         {
           handle->info->aordblks++;
           handle->info->uordblks += nodesize;
         }
+#else
+      if (handle->dump->pid == MM_BACKTRACE_ALLOC_PID ||
+          handle->dump->pid == node->pid ||
+          (handle->dump->pid == MM_BACKTRACE_INVALID_PID &&
+            !nxsched_get_tcb(node->pid)))
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_mallinfo.c:109]:
```suggestion
           nxsched_get_tcb(node->pid) == NULL))
```

File: mm/mm_heap/mm_mallinfo.c [hunk 3]
```diff
@@ -98,16 +98,24 @@ static void mallinfo_task_handler(FAR struct mm_allocnode_s *node,
...
+        {
+          if (node->seqno >= handle->dump->seqmin &&
+                  node->seqno <= handle->dump->seqmax)
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_mallinfo.c:112]:
```suggestion
          if (node->seqno >= handle->dump->seqmin &&
              node->seqno <= handle->dump->seqmax)
```

**Review by jerpelea — APPROVED** [review]

File: mm/mm_heap/mm_malloc.c [hunk 4]
```diff
@@ -81,12 +81,12 @@ void mm_dump_handler(FAR struct tcb_s *tcb, FAR void *arg)
...
   struct mm_memdump_s dump;
 
-  dump.pid = tcb->pid;
+  dump.pid = tcb ? tcb->pid : MM_BACKTRACE_INVALID_PID;
   dump.seqmin = 0;
   dump.seqmax = ULONG_MAX;
   info = mm_mallinfo_task(arg, &dump);
   mwarn("pid:%5d, used:%10d, nused:%10d\n",
-        tcb->pid, info.uordblks, info.aordblks);
+        tcb ? tcb->pid : -1, info.uordblks, info.aordblks);
```
Comment (pkarashchenko) [inline mm/mm_heap/mm_malloc.c:89]:
```suggestion
       dump.pid, info.uordblks, info.aordblks);
```

File: mm/tlsf/mm_tlsf.c [hunk 5]
```diff
@@ -294,16 +294,24 @@ static void mallinfo_task_handler(FAR void *ptr, size_t size, int used,
...
-#endif
         {
           handler->info->aordblks++;
           handler->info->uordblks += size;
         }
+#else
+      if (handler->dump->pid == MM_BACKTRACE_ALLOC_PID ||
+          handler->dump->pid == dump->pid ||
+          (handler->dump->pid == MM_BACKTRACE_INVALID_PID &&
+            !nxsched_get_tcb(dump->pid)))
```
Comment (pkarashchenko) [inline mm/tlsf/mm_tlsf.c:305]:
```suggestion
           nxsched_get_tcb(dump->pid) == NULL))
```

File: mm/tlsf/mm_tlsf.c [hunk 6]
```diff
@@ -294,16 +294,24 @@ static void mallinfo_task_handler(FAR void *ptr, size_t size, int used,
...
+        {
+          if (dump->seqno >= handler->dump->seqmin &&
+                  dump->seqno <= handler->dump->seqmax)
```
Comment (pkarashchenko) [inline mm/tlsf/mm_tlsf.c:308]:
```suggestion
          if (dump->seqno >= handler->dump->seqmin &&
              dump->seqno <= handler->dump->seqmax)
```

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by patacongo** [comment]
@xiaoxiang781216 The name PID is used throughout the change.  But isn't this actually the thread ID (TID) and not the process ID (PID).  dump->pid comes from tcb->pid.  But that field is misnamed, tcb->pid is really the TID and should be renamed tcb->tid.  It is the value that is returned when gettid() is called.

If you call getpid(), you get tcb->group->tg_pid.  That is correctly the PID.

**Comment by xiaoxiang781216** [comment]
@patacongo, @acassis just create https://github.com/apache/nuttx/issues/9421 to track this problem. Let's fix it step by step.

**Pull request merged**